docs/src/capi/explanation/in-place-upgrades.md

```diff
@@ -0,0 +1,160 @@
+trigger an in-place upgrade for that machine. The single machine upgrader 
+(which is watching for changes on machines), notices this annotation 
+and attempts to upgrade the Kubernetes version of that machine to the 
+specified version.
+
+Upgrade methods or options can be specified to upgrade to a snap channel, 
+revision, or install a new snap from a file already placed on the 
+machine to make up for air-gapped environments.
+
+A successfully upgraded machine shows the following annotations:
+
+```yaml
+annotations:
+  v1beta2.k8sd.io/in-place-upgrade-release: "channel=1.31/stable"
+  v1beta2.k8sd.io/in-place-upgrade-status: "done"
+```
+
+If the upgrade fails, the controller will mark the machine and retry 
+the upgrade immediately:
+
+```yaml
+annotations:
+  # the `upgrade-to` causes the retry to happen
+  v1beta2.k8sd.io/in-place-upgrade-to: "channel=1.31/stable"
+  v1beta2.k8sd.io/in-place-upgrade-status: "failed"
+
+  # orchestrator will notice this annotation and knows that the 
+  # upgrade for this machine failed
+  v1beta2.k8sd.io/in-place-upgrade-last-failed-attempt-at: "Sat, 7 Nov 
+  2024 13:30:00 +0400"
+```
+
+For the detailed implementation of this controller, make sure to 
+checkout the [{{product}} CAPI repository][capi-repo].
+
+Overall, by applying and removing annotations, the single machine 
+upgrader determines the upgrade status of the machine it’s trying to 
+reconcile and takes necessary actions to successfully complete an 
+in-place upgrade. The following diagram shows the flow of the in-place 
+upgrade of a single machine:
+
+![Diagram][img-single-machine]
+
+### Machine Upgrade Process
+
+{{product}} has a daemon running in the background called the `k8sd`. 
+It’s responsible for many things in the context of {{product}} but 
+one of them is to expose certain endpoints that can be used to 
+interact with the cluster. The single machine upgrader calls the 
+`/snap/refresh` endpoint on the machine that it’s trying to upgrade. 
+This endpoint call will trigger the “actual” upgrade process and in 
+the meantime, the `/snap/refresh-status` will be called periodically 
+by the single machine upgrader to see how things are going. It’s worth 
+noting that ensuring secure communication between the single machine 
+upgrader and the {{product}} daemon (k8sd) is really important and 
+something that is handled internally as well.
+
+![Diagram][img-k8sd-call]
+
+### In-place upgrades on large workload clusters
+
+While the “Single Machine In-Place Upgrade Controller” is responsible 
+for upgrading individual machines, the "Orchestrated In-Place Upgrade 
+Controller" makes sure that groups of machines will get upgraded.
+By applying the `upgrade-to` annotation on an object that owns machines 
+(e.g. a `MachineDeployment`), this controller will mark the owned machines 
+one by one which will cause the "Single Machine Upgrader" to pickup those 
+annotations and upgrade the machines.
```

Comment:
```suggestion
annotations and upgrade the machines. To avoid undesirable situations like quorum loss or severe downtime, these upgrades happen in sequence. 
```